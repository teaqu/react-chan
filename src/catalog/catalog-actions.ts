import { Catalog } from './catalog';
import { Thread } from './thread';
import { ThunkAction } from 'redux-thunk';
import { RootState } from 'src/shared/root-reducer';

type ThunkResult<R> = ThunkAction<R, RootState, undefined, CatalogActionTypes>;

export const REQUEST_CATALOG = 'REQUEST_CATALOG';
export interface RequestCatalogAction {
  type: typeof REQUEST_CATALOG;
  board: string;
}
function requestCatalog(board: string): RequestCatalogAction {
  return {
    type: REQUEST_CATALOG,
    board
  };
}

export const RECEIVE_CATALOG = 'RECEIVE_CATALOG';
export interface ReciveCatalogAction {
  type: typeof RECEIVE_CATALOG;
  board: string;
  threads: Thread[];
  receivedAt: number;
}
function recieveCatalog(board: string, threads: Thread[]): ReciveCatalogAction {
  return {
    type: RECEIVE_CATALOG,
    board,
    threads: threads,
    receivedAt: Date.now()
  };
}

export function fetchCatalog(board: string): ThunkResult<void> {
  return function(dispatch) {
    dispatch(requestCatalog(board));
    return fetch(`https://a.4cdn.org/${board}/catalog.json`)
      .then(
        response => response.json(),
        error => console.log('An error occurred.', error)
      )
      .then((catalogs: Catalog[]) => {
        const threads = catalogs
          .map(catalog => {
            return catalog.threads.map(thread => {
              thread.page = catalog.page;
              return thread;
            });
          })
          .reduce((prev, curr) => {
            return prev.concat(curr);
          });
        dispatch(recieveCatalog(board, threads));
      });
  };
}

function shouldFetchCatalog(state: RootState) {
  const catalog = state.catalog;
  if (!catalog) {
    return true;
  } else if (catalog.isFetching) {
    return false;
  } else {
    return true;
  }
}

export function fetchCatalogIfNeeded(board: string): ThunkResult<void> {
  return (dispatch, getState) => {
    if (shouldFetchCatalog(getState())) {
      return dispatch(fetchCatalog(board));
    }
  };
}

export type CatalogActionTypes = RequestCatalogAction | ReciveCatalogAction;
