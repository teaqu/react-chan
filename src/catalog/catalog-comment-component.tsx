import React, { ReactNode } from 'react';
import HTML, {
  HtmlAttributesDictionary,
  RendererDictionary,
  PassProps,
  NonRegisteredStylesProp
} from 'react-native-render-html';
import { StyleSheet } from 'react-native';

import { SpoilerComponent } from '../comment/spoiler-component';

type Props = { comment: string };
export const CatalogCommentComponent = (props: Props) => {
  const renderers = (): RendererDictionary => {
    return {
      s: {
        renderer: (
          htmlAttribs: HtmlAttributesDictionary,
          children: ReactNode,
          convertedCSSStyles: NonRegisteredStylesProp<any>,
          passProps: PassProps
        ) => <SpoilerComponent children={children} key={passProps.key} />,
        wrapper: 'Text'
      }
    };
  };

  // Remove unnecessary whitespace
  const comment = props.comment.replace('<br><br>', '<br>');

  return (
    <HTML
      html={'<p>' + comment + '</p>'}
      renderers={renderers()}
      textSelectable={true}
      tagsStyles={{
        p: {
          textAlign: 'center'
        },
        s: {
          textDecorationLine: 'none',
          backgroundColor: 'black'
        }
      }}
      containerStyle={styles.com_container}
    />
  );
};

const styles = StyleSheet.create({
  com_container: {
    maxHeight: 100,
    overflow: 'hidden'
  }
});
