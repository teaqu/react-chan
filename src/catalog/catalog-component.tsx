import React, { useEffect } from 'react';
import { FlatList, StyleSheet, RefreshControl } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { RootState } from 'src/shared/root-reducer';

import { Thread } from '../thread/thread';

import { CatalogThreadComponent } from './catalog-thread-component';
import actions from './catalog-actions';

export function CatalogComponent() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(actions.fetchCatalog('a'));
  }, [dispatch]);

  const threads: Thread[] = useSelector(
    (state: RootState) => state.catalog.threads
  );
  const isFetching: boolean = useSelector(
    (state: RootState) => state.catalog.isFetching
  );

  const onRefresh = React.useCallback(() => {
    dispatch(actions.fetchCatalog('a'));
  }, [dispatch]);

  const renderItem = (item: any) => {
    return <CatalogThreadComponent thread={item.item} />;
  };

  const keyExtractor = (item: any) => {
    return item.no.toString();
  };

  return (
    <FlatList<Thread>
      refreshControl={
        <RefreshControl refreshing={isFetching} onRefresh={onRefresh} />
      }
      style={styles.catalog}
      data={threads}
      numColumns={3}
      keyExtractor={keyExtractor}
      renderItem={renderItem}
    />
  );
}

const styles = StyleSheet.create({
  catalog: {
    backgroundColor: '#eef2ff'
  }
});
