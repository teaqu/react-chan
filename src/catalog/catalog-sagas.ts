import { call, put, takeLatest, getContext } from 'redux-saga/effects';
import { AnyAction } from 'redux';

import { Thread } from 'src/thread/thread';
import { ChanAPI } from 'src/shared/chan-api/chan-api';

import actions from './catalog-actions';

export function* fetchCatalog(action: AnyAction) {
  try {
    const chanAPI: ChanAPI = yield getContext('chanAPI');
    const threads: Thread[] = yield call(chanAPI.fetchCatalog, action.payload);
    yield put(actions.fetchCatalogSucceeded(action.payload, threads));
  } catch (e) {
    yield put(actions.fetchCatalogFailed(e.message));
  }
}

export function* watchFetchCatalog() {
  yield takeLatest(actions.fetchCatalog, fetchCatalog);
}
