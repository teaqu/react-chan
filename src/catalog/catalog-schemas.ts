import { schema } from 'normalizr';

const threadSchema = new schema.Entity(
  'threads',
  {},
  {
    idAttribute: 'no',
    processStrategy: (entity, parent) => {
      return { ...entity, page: parent.page };
    }
  }
);
export const catalogSchema = new schema.Entity(
  'catalog',
  {
    threads: [threadSchema]
  },
  { idAttribute: 'page' }
);
