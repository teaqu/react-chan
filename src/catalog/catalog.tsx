import { Thread } from "src/thread/thread";

export interface Catalog {
    page: number;
    threads: Thread[];
}