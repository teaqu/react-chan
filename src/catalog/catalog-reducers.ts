import { createReducer } from '@reduxjs/toolkit';
import { Thread } from './thread';
import { requestCatalog, recieveCatalog } from './catalog-actions';

export interface CatalogState {
  isFetching: boolean;
  threads: Thread[];
}

const initialState: CatalogState = {
  isFetching: false,
  threads: []
};

export default createReducer(initialState, {
  [requestCatalog.type]: state => {
    state.isFetching = true;
  },
  [recieveCatalog.type]: (state, action) => {
    state.isFetching = false;
    state.threads = action.payload.threads;
  }
});
