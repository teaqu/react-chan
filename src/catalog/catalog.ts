import { Thread } from '@/catalog/thread';

export interface Catalog {
  page: number;
  threads: Thread[];
}
