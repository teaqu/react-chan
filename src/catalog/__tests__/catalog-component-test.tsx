import React from 'react';
import { render, cleanup } from 'react-native-testing-library';

import { CatalogComponent } from 'src/catalog/catalog-component';
import actions from 'src/catalog/catalog-actions';

const mockedSelector = jest.fn();
const mockedDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: () => mockedSelector,
  useDispatch: () => mockedDispatch
}));

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn()
  },
  ...props
});

describe('catalog component', () => {
  const props: any = createTestProps({
    catalog: {
      isFetching: true,
      threads: []
    }
  });

  const catalogComponent = render(<CatalogComponent {...props} />);

  it('renders correctly', () => {
    expect(catalogComponent.toJSON()).toMatchSnapshot();
    expect(mockedDispatch).toHaveBeenCalledWith(actions.fetchCatalog('a'));
  });

  afterEach(cleanup);
});
