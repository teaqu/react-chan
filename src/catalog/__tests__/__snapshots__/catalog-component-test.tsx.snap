// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalog component renders correctly 1`] = `
<RCTScrollView
  data={[MockFunction]}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={3}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={[MockFunction]}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "backgroundColor": "#eef2ff",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View />
</RCTScrollView>
`;
